### PR TITLE
Add Support for `Async` openai instrumentation

### PR DESCRIPTION
**Pull request by alizenhom**
# Description
- Add wrapper around `AsyncCompletions.create` to instrument async openai functions
- Adjust `StreamWrapper` to include `__anext__`, `__aiter__` & `__aenter__`

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [x] `tox -e py312-test-instrumentation-openai-v2-0`
- [x] `tox -e py312-test-instrumentation-openai-v2-1`

# Does This PR Require a Core Repo Change?
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_404.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      - 1.26.0
+      x-stainless-runtime:
+      - CPython
+      x-stainless-runtime-version:
+      - 3.12.5
+    method: POST
+    uri: https://api.openai.com/v1/chat/completions
+  response:
+    body:
+      string: "{\n    \"error\": {\n        \"message\": \"The model `this-model-does-not-exist`
```
Comment (xrmx) [inline instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_404.yaml:38]: Have you recorded this before recent changes to conftest?

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_404.yaml:38]: opps, will adjust now

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py [hunk 2]
```diff
@@ -52,7 +52,7 @@
 from opentelemetry.semconv.schemas import Schemas
```
Comment (gyliu513) [inline instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py:52]: - Can you update changelog as well https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md
- In README, seems the link [OpenTelemetry OpenAI Instrumentation](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/openai/openai.html) is not working any more.
- Can we add a list of instrumented APIs for OpenAI in readme as well?

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py:52]: sure!

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py:52]: @gyliu513, added changelog and adjusted readme

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/__init__.py:52]: reverted changes, as i see @lmolkova having the same changes specified in this [PR](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2988)

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+- Added a wrapper for `AsyncCompletions.create` inside `src/opentelemetry/instrumentation/openai_v2/__init__.py` to instrument async chat completions
```
Comment (codefromthecrypt) [inline instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md:10]: is this typical amount of detail we want in the release notes? I was thinking that from an end user viewpoint, the main change is support for `AsyncOpenAI`/`AsyncCompletions`. The various steps to implement this, might be considered too much information, especially if we repeat this pattern for every update to this package. Thoughts?

**Review by codefromthecrypt — APPROVED** [review]
(non-binding approve)

Looks good with a suggestion on changelog, and assuming the recordings are corrected to the formatting of the others.

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_404.yaml (hunk 1 above)
Comment (codefromthecrypt) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_404.yaml:38]: I think this is the problem, maybe someone can add the skip-changelog tag on it and merge. Then, this PR can be rebased over it. https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2989

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md (hunk 3 above)
Comment (xrmx) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md:10]: Yes!

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md:10]: noted, will adjust

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_tool_calls_no_content.yaml [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+      - application/json
+      accept-encoding:
+      - gzip, deflate
+      connection:
+      - keep-alive
+      content-length:
+      - '746'
+      content-type:
+      - application/json
+      cookie:
```
Comment (codefromthecrypt) [inline instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_tool_calls_no_content.yaml:231]: I know this is annoying to revise this change again, but I looked more carefully this time, and noticed the leaked cookie. We should prevent this always and so I made a different PR. Once that's in, you should rebase this, delete the test data around tool calls at least then run again.

https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2993

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_tool_calls_no_content.yaml:231]: No worries, acknowledged @codefromthecrypt

Comment (alizenhom) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/cassettes/test_async_chat_completion_tool_calls_no_content.yaml:235]: @codefromthecrypt rebased, re-recorded tests & scrubbed cookies successfully

**Review by codefromthecrypt — APPROVED** [review]

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+- Support for `AsyncOpenAI/AsyncCompletions`
```
Comment (lzchen) [inline instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md:10]: Nit: Please use the similar format for CHANGELOG as the other entries

**Review by lzchen — APPROVED** [review]

File: instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md (hunk 5 above)
Comment (codefromthecrypt) [reply instrumentation-genai/opentelemetry-instrumentation-openai-v2/CHANGELOG.md:10]: @alizenhom I think this means put the PR here in parens like those below.

**Review by xrmx — APPROVED** [review]

**Review by karthikscale3 — APPROVED** [review]

**Pull request merged**